src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs

```diff
@@ -0,0 +1,1504 @@
+        let bucketSize = bucketSizeFromBitDepth bitdepth
+        let bucketBitMask = calcSubBitMask bitdepth
+        let newBucket = Array.zeroCreate<Shard<'K,'V>>(bucketSize)
+
+        items
+        |> Seq.iter (fun item ->
+            let kvp =  mapFn item
+            let bi, si = getIndexes(kvp.Key,bucketBitMask)
+            // let kh =   kvp.Key.GetHashCode()
+            // let bi =   util.bucketIndex(kh,bucketBitMask)
+            // let si =   util.shardIndex(kh)
+            let shrd = util.getOrCreateShard(newBucket,bi)
+            let m =    shrd.[si]
+            shrd.[si] <- if isNull m then MapOne (kvp.Key,kvp.Value) else MapTree.add comparer kvp.Key kvp.Value m
+            )
+        
+        ShardMap<'K,'V>(counter,newBucket)
+
+    static member OfKeyedList (keyFn:'V -> 'K) (items:'V list) = ShardMap<'K,'V>.init(List.length items,items,fun v -> KeyValuePair<'K,'V>(keyFn v,v))
+        
+    static member OfList (items:('K * 'V) list) = ShardMap<'K,'V>(List.length items,items)
+        
+    static member OfSeq (keyFn:'V -> 'K) (keyValueTuples:'V seq) = 
+        let mutable counter = 0
+        let mutable items = []
+        keyValueTuples |> Seq.iter (fun kvp -> 
+            counter <- counter + 1
+            items <- kvp :: items
+        )
+        ShardMap<'K,'V>.init<_>(counter,items,fun v -> KeyValuePair<'K,'V>(keyFn v,v))
+
+    static member OfArray(keyValueTuples:('K * 'V) []) = 
+        ShardMap<'K,'V>.init<_>(keyValueTuples.Length,keyValueTuples,fun (k,v) -> KeyValuePair<'K,'V>(k,v))
+
+    static member OfListMapped  (fn:'T -> KeyValuePair<'K,'V>)  (items:'T list) = ShardMap<'K,'V>.init<_>(items.Length,items,fn)
+    static member fold f s (m:ShardMap<_,_>) = m.Fold f s
+
+    ////////////////////////////////////
+    /// Contructors
+    ///////////////////////////////////
+    
+    new(capacity:int) =
+        let size = if capacity < ShardSize then ShardSize else capacity
+        let bitdepth = calcBitMaskDepth size
+        let bucketSize = bucketSizeFromBitDepth bitdepth
+        let newBucket = Array.zeroCreate<Shard<'K,'V>>(bucketSize)
+        ShardMap<_,_>(0,newBucket)
+
+    new(count:int,kvps:('K * 'V) list) = let count,bucket = init<_,'K,'V>(count, kvps ,fun (k,v) -> KeyValuePair<'K,'V>(k,v)) in ShardMap<'K,'V>(count,bucket)
+    new(kvps:('K * 'V) list) = ShardMap<'K,'V>(List.length kvps,kvps)
+
+    new(kvps:('K * 'V) [])   = let count,bucket = init<_,'K,'V>(kvps.Length, kvps ,fun (k,v) -> KeyValuePair<'K,'V>(k,v)) in ShardMap<'K,'V>(count,bucket)
+
+    new(count: int, kvps:('K * 'V) seq)  = 
+        let count,bucket = init<_,'K,'V>(count, kvps ,fun (k,v) -> KeyValuePair<'K,'V>(k,v)) in ShardMap<'K,'V>(count,bucket)
+    new(kvps:('K * 'V) seq)  = 
+        let mutable counter = 0
+        let mutable items = []
+        kvps |> Seq.iter (fun kvp -> 
+            counter <- counter + 1
+            items <- kvp :: items
+        )
+        ShardMap<'K,'V>(counter,kvps)
```

Comment:
items is unused after being initialized?

Reply:
apologies, this still worked due to re-running seq iterator but the whole point was to run only once with count to pre-compute the hash space needed, I have updated in source as was an oversight.